### PR TITLE
chore(ci): CI/Docker hardening — multi-stage backend image, Node 20, pinned uv, isort (#176)

### PR DESCRIPTION
**Pull request by frankbria**
## Summary

Implements the **mechanical subset** of the #150-review CI/Docker follow-ups tracked in #176. Scope was confirmed with the maintainer: the large/long-term items (mypy→blocking, broaden integration coverage) stay open; pyupgrade churn is excluded.

| # | Item | Done |
|---|------|------|
| 2 | Backend Dockerfile → multi-stage | ✅ |
| 3 | Node 18 → 20 (workflows + frontend Dockerfile) | ✅ |
| 4 | Pin `uv` image (`:0.5` → `:0.5.31@sha256:…`) | ✅ |
| 6 | Drop `npm ci \|\| npm install` fallback | ✅ |
| 5 | Ruff: enable more rules | ✅ (import-sorting only) |

### Details
- **Backend Dockerfile multi-stage** — `build-essential` (~200 MB, needed only to compile wheels during `uv sync`) is now confined to a `builder` stage. The slim `runtime` stage copies only the built `/app` (virtualenv + source) and installs just `curl` (healthcheck) + `libgomp1` (xgboost/lightgbm). Both stages share `python:3.13-slim` so the interpreter path baked into `/app/.venv` stays valid.
- **Pin uv** — `ghcr.io/astral-sh/uv:0.5` floated within `0.5.x`; now pinned to `0.5.31@sha256:7bff3c37…` (patch tag + digest) for reproducible image builds.
- **Node 18 → 20** — Node 18 is EOL. Bumped all 6 workflow pins (`ci.yml` ×3, `e2e-tests.yml`, `smoke-tests.yml`, `perf-tests.yml`) and the 3 `FROM node:` lines in `apps/frontend/Dockerfile`.
- **Drop npm fallback** — `npm ci || (… npm install)` masked lockfile drift; replaced with plain `npm ci` in `ci.yml` (×3) and `e2e-tests.yml`. Lockfile verified in sync (`npm ci --dry-run` clean).
- **Ruff import-sorting** — `extend-select = ["I"]` added on top of the default rule set; 314 `I001` findings auto-fixed. `app/main.py` is `I001`-ignored to preserve its deliberate sys.path-mutation import order (#149).

## Verification
- `docker build` of the backend image **succeeds**; inside the runtime image: `import app.main` + `xgboost`/`lightgbm` succeed, **`gcc` absent** (build-essential not carried forward), `curl` present.
- `ruff check .` — clean.
- Service-free backend pytest suite — green (imports intact after the reorder; `app.main` imports cleanly).
- All workflow YAML parses; no remaining Node 18 / `npm ci ||` / bare `uv:0.5` references.

## Known Limitations / Deferred (issue #176 stays open)
- **#1 mypy → blocking** — ~317 findings; an incremental typing project, not a single PR.
- **#7 broaden integration coverage** — needs reliable LocalStack/OpenAI provisioning to avoid an environment-flaky required gate.
- **ruff pyupgrade (UP\*)** — ~2,900 findings (repo-wide annotation rewrites) that would collide with the mypy typing effort; deliberately not enabled.
- The frontend Node-20 bump and backend image are exercised in CI here; full container runtime (gunicorn + Mongo) is covered by the staging deploy, not this PR.

Closes part of #176 (mechanical items 2–6); leaves #1, #7, and pyupgrade tracked.

## Release Notes

* **Infrastructure & Build Improvements**
  * Updated Node.js from 18 to 20 across CI workflows and the frontend Docker image.
  * Standardized dependency installs to use `npm ci` only (no install fallback).
  * Improved smoke-test execution by increasing setup timeout and caching Playwright’s Chromium downloads.
  * Updated the backend production image to use digest-pinned Python 3.13-slim with a smaller multi-stage build.

* **Code Quality**
  * Extended Ruff import-sorting checks to enforce consistent import ordering.

### CODE REVIEW

File: apps/backend/app/models/batch_job.py [hunk 1]
```diff
@@ -2,11 +2,12 @@
...
 
-from typing import List, Dict, Any, Optional
 from datetime import datetime
-from beanie import Document, Indexed
-from pydantic import Field, BaseModel
 from enum import Enum
+from typing import Any, Dict, List, Optional
+
+from beanie import Document, Indexed
+from pydantic import BaseModel, Field
```
[the remark this answers is not in the document]
Comment (frankbria) [reply apps/backend/app/models/batch_job.py]: Addressed in da607d8. `timedelta` is now in the top `from datetime import datetime, timedelta` block and the module-bottom import is removed. Since that was the only post-code import in the file, I also dropped the now-unnecessary `app/models/batch_job.py` E402 per-file-ignore. (Note: it was not actually a runtime `NameError` hazard — module-bottom imports execute at import time, before any property is ever called — but it's genuine import hygiene and on-theme for this PR.) ruff clean; model tests green.

File: apps/backend/tests/test_api/test_visualizations.py [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
+
+from app.auth.nextauth_auth import get_current_user_id
+from app.main import app
+from app.models.user_data import UserData
 from app.models.visualization_cache import (
-    HistogramData,
     BoxplotData,
     CorrelationMatrixData,
+    HistogramData,
 )
```
[the remark this answers is not in the document]
Comment (frankbria) [reply apps/backend/tests/test_api/test_visualizations.py:14]: Skipping in this PR (with reasoning). This is pre-existing test code — this PR's only change to the file is isort reordering the top import block (lines 1–14); the `Link(dataset.id, document_class=UserData)` call at line 67 is untouched. `dataset` is a `MagicMock(spec=UserData)` and `.link` is just a fixture attribute; the construction works under the pinned Beanie version (the suite is green), so this is a best-practice critique of a passing test rather than a defect introduced here. Per the project's verify-before-fix / anti-test-mutation rule, I won't rewrite a passing test fixture on a style basis inside a CI/Docker hardening PR. Worth a dedicated test-hygiene follow-up if desired.

File: .github/workflows/smoke-tests.yml [hunk 3]
```diff
@@ -51,6 +56,18 @@ jobs:
...
+      # setup cost. Keyed on the lockfile so a Playwright version bump rebuilds
+      # it. `--with-deps` still runs each time for the apt system libraries
+      # (not stored in this cache path), but those are mostly preinstalled.
+      - name: Cache Playwright browsers
+        uses: actions/cache@v4
+        with:
+          path: ~/.cache/ms-playwright
+          key: ${{ runner.os }}-playwright-${{ hashFiles('apps/frontend/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.os }}-playwright-
```
[the remark this answers is not in the document]
Comment (frankbria) [reply .github/workflows/smoke-tests.yml:69]: Valid supply-chain point, but deferring (with reasoning). Every workflow in this repo pins actions by version tag (`@v4`/`@v5`), not commit SHA. SHA-pinning only the one `actions/cache@v4` I just added would create an inconsistent island; doing it properly means pinning all actions across all workflows (and ideally the CI service-container images: mongo/redis/localstack/minio) as one repo-wide policy. That's a separate hardening task — tracked as a follow-up on #176, not folded into this CI/Docker PR.
